### PR TITLE
refactored ceos certificate generation and config push with scrapligo

### PR DESCRIPTION
**Pull request by hellt**
Hi @marcushines 
I wanted to propose to change the way kne works with CLI based provisioning over `kubectl exec`.
Namely, I propose to use the [scrapligo](https://github.com/scrapli/scrapligo) package to interact with networking devices over a PTY that `kubectl exec` offers.

As part of this PR you will find the working implementation of the said change for ceos certs generation.

The benefits of using scrapligo:

1. proven prompt and privilege level escalation handling. scrapli and scrapligo are used widely in the network automation sphere, thus it has a proven prompt detection and privilege level handling for major OSes.
2. the internal mechanics of priv level escalations simplifies the code as we do not need to worry about all the spawner things and timeout handling, it is handled in scrapligo
3. by using this package we add support for many other OSes without polluting kne with those details

As far as SR Linux is concerned, I plan to use scrapligo for self signed certs generation as part of the controller operation routine.

Let me know if you find those benefits compelling.

PS. the config push as well can leverage scrapligo and this will even further make the kne more lightweight and robust.
PS2. the tests are expectedly failing, as I didn't remove the tests for fakeSpawner

### CODE REVIEW

**Comment by hellt** [comment]
For some reason the pipeline picked the wrong commit for build/tests

@marcushines can you please re trigger it?

**Comment by hellt** [comment]
@carlmontanari I've added some makeup to this, namely I fixed the pod name to be `pod1` as that is what patched k8s client expected to see instead of testEosNode.

I think the reason CI is picking old commit for testing is because the CLA is not signed...

File: topo/node/ceos/ceos.go [hunk 1]
```diff
@@ -30,24 +39,83 @@ func New(pb *topopb.Node) (node.Implementation, error) {
...
-	return expect.Spawn(command, timeout, opts...)
+	return nil
 }
 
-var (
-	timeSecond = time.Second
-)
+// PatchCLIConnOpen sets the OpenCmd and ExecCmd of system transport to work with `kubectl exec` terminal.
+func (n *Node) PatchCLIConnOpen(ns string) error {
+	tIntf := reflect.ValueOf(n.cliConn.Transport)
```
[the remark this answers is not in the document]
Comment (mhines01) [reply topo/node/ceos/ceos.go:69]: so this looks like you are just getting Transport that is part of the base.Driver?

// PatchCLIConnOpen sets the OpenCmd and ExecCmd of system transport to work with `kubectl exec` terminal.
func (n *Node) PatchCLIConnOpen(ns string) error {
	t, ok := n.cliConn.Driver.Transport.(*scraplitransport.System)
	if !ok {
		return fmt.Errorf("invalid underlying transport type: %T", n.cliConn.Driver.Transport)
	}
	t.ExecCmd = "kubectl"
	t.OpenCmd = []string{"exec", "-it", "-n", ns, n.pb.Name, "--", "Cli"}
	return nil
}

**Comment by carlmontanari** [comment]
@googlebot I fixed it.

File: topo/node/ceos/ceos.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlmontanari) [reply topo/node/ceos/ceos.go:69]: > so this looks like you are just getting Transport that is part of the base.Driver?

Yeah -- scrapli has multiple transport options so we have to check the type... moreover we have to check for the *fields* because the mock test transport (that patches the reads/writes w/ the stored "session file") only embeds the transport type that we use here (`system` transport -- this is the subprocess wrapper one basically). Outside of the testing use case this could just type assert that `Transport` is of type `transport.System` but the testing bit adds a bit of fun to this.

So TL;DR we have to check if the transport has the `ExecCmd`/`OpenCmd` field attached to it.

[the remark this answers is not in the document]
Comment (hellt) [reply topo/node/ceos/ceos.go:69]: My 2c is that here the usage of reflect is somewhat innocent. 

We merely set the two fields of the underlying struct to make the testing work on the whole Generate SelfSigned function instead of the underlying scrapligo pieces

[the remark this answers is not in the document]
Comment (hellt) [reply topo/node/ceos/ceos.go:69]: Also tests are red because CI is picking not the tip of the branch for whatever reason

[the remark this answers is not in the document]
Comment (marcushines) [reply topo/node/ceos/ceos.go:69]: you need to rebase this PR against head

[the remark this answers is not in the document]
Comment (marcushines) [reply topo/node/ceos/ceos.go:69]: using reflect in production code should almost always be avoided...
If you need to access a lower struct member then - make the member public / provide an accessor 
in this case you just need a type assertion on an interface

[the remark this answers is not in the document]
Comment (hellt) [reply topo/node/ceos/ceos.go]: Done. Thank you

File: topo/node/ceos/ceos.go [hunk 2]
```diff
@@ -83,127 +149,95 @@ func (n *Node) GenerateSelfSigned(ctx context.Context, ni node.Interface) error
...
+			selfSigned.KeyName,
+			n.pb.Name,
+		),
 	}
-	_, _, err = g.Expect(regexp.MustCompile(`(.*)#`), 30*timeSecond)
+
+	r, err := n.cliConn.SendConfigs(cfgs)
 	if err != nil {
 		return err
+	} else if r.Failed() {
```
[the remark this answers is not in the document]
Comment (mhines01) [reply topo/node/ceos/ceos.go:177]: Also i would question why is the this the API pattern?

Either "sendconfigs" works it doesn't

if you want to have a different action based on the type of error then build that into the api

https://play.golang.org/p/P9LcmbJna6M

basically the go style would say if you get an err you cannot trust the return value.

there are a couple examples in go there that is done - for one the file io functions do return both return values and err's specifically for handling io.EOF but that is really a pattern based on the OS semantics not on it being a good patten

File: topo/node/ceos/ceos.go [hunk 3]
```diff
@@ -83,127 +149,95 @@ func (n *Node) GenerateSelfSigned(ctx context.Context, ni node.Interface) error
...
-		return err
-	}
-	_, _, err = g.Expect(regexp.MustCompile(`#`), -1)
+
+	defer n.cliConn.Close()
+
+	resp, err := n.cliConn.SendConfig(cfgs)
 	if err != nil {
 		return err
+	} else if resp.Failed {
```
[the remark this answers is not in the document]
Comment (hellt) [reply topo/node/ceos/ceos.go]: addressed in eca3d33

File: topo/node/ceos/ceos.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hellt) [reply topo/node/ceos/ceos.go]: Hi @marcushines 
`else if` addressed in eca3d33

the decision to have `r.Failed()` for multiline commands and `r.Failed` for single commands was deliberate. It indeed might look weird, but that is because it is not an error in the context of the `SendXXX` operation.

The way scrapligo treats operations over SSH/TTY transports and its errors is the following:

1. when a user attempts to send a CLI command towards a device it uses multiple vehicles exposed in scrapligo api: `channel`, `session`, `transport` layers .
2. any error that might appear on the above  layers will return an error if things went utterly bad and the attempt to **send** the command failed.
3. if the low level things worked out and the user was able to send the command, scrapligo marks this as a successful operation and the error is `nil`.
4. If a user sends a command such as `**^%*&%` and scrapli detects that this command was received by a device and a device returned expected prompt back, this is a successful operation.
5. Since the act of sending the command/config was successful, we have a `nil` error. It is very much like in HTTP libraries you may succeed in sending HTTP request but you check the status (i.e. for `404`).
6. And that _Status_ is what `Failed` is. It will simply indicate if scrapli detected that whatever the user sent over was not accepted by the Network OS (still the send operation was totally good) application wise.

I hope this makes sense :)

[the remark this answers is not in the document]
Comment (hellt) [reply topo/node/ceos/ceos.go]: Maybe a practical use cases that I just had with ceos that explains it even more precise.

I had the following set of set configs:

```
cfgs := []string{
		"no ip address",
		"no ipv6 address",
		"interface management 0",
	}
```

I sent this with scrapligo `SendConfigs` and what scrapligo did is happily sent it over, declared things were good.
They were good, because scrapligo sent the commands over, they were processed by the CLI and on the transport side of things it all checked out. The commands were succesfully sent and they were applied in the correct privilege level.

The problem is that CLI of EOS emitted some errors, because my order of operations was incorrect. To indicate that there was an application layer issue, scrapligo uses `Failed` flag on the response object.

**Review by marcushines — APPROVED** [review]
